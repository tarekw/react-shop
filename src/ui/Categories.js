import React from "react";
import PropTypes from "prop-types";

import ListGroup from "react-bootstrap/ListGroup";

export const Categories = ({
  categories,
  selectedIndex = 0,
  setSelectedIndex = (f) => f,
}) => (
  <ListGroup as="ul">
    {categories.map((item, index) => (
      <ListGroup.Item
        onClick={() => {
          setSelectedIndex(index);
        }}
        active={selectedIndex === index}
        key={index} // shouldn't use index as keys, but we don't intend to change the order of items here
        style={{ cursor: "pointer" }}
      >
        {item}
      </ListGroup.Item>
    ))}
  </ListGroup>
);

Categories.propTypes = {
  categories: PropTypes.array,
  selectedIndex: PropTypes.number,
  setSelectedIndex: PropTypes.func,
};

export default React.memo(Categories);
