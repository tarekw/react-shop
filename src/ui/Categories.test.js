import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

import ListGroup from "react-bootstrap/ListGroup";

import { Categories } from "./Categories";

describe("Categories", () => {
  let categories;
  let setSelectedIndex;

  beforeEach(() => {
    categories = [
      "electronics",
      "jewelery",
      "men's clothing",
      "women's clothing",
    ];
    setSelectedIndex = jest.fn();
  });

  it("should render properly", () => {
    const tree = renderer
      .create(
        <Categories
          categories={categories}
          setSelectedIndex={setSelectedIndex}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should handle item selection", () => {
    const wrapper = shallow(
      <Categories categories={categories} setSelectedIndex={setSelectedIndex} />
    );
    wrapper.find(ListGroup.Item).last().simulate("click");
    expect(setSelectedIndex).toHaveBeenCalledTimes(1);
    expect(setSelectedIndex).toHaveBeenCalledWith(3);
  });
});
