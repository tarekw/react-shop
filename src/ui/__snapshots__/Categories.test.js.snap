// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories should render properly 1`] = `
<ul
  className="list-group"
  onKeyDown={[Function]}
>
  <div
    className="list-group-item active"
    data-rr-ui-event-key={null}
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    electronics
  </div>
  <div
    className="list-group-item"
    data-rr-ui-event-key={null}
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    jewelery
  </div>
  <div
    className="list-group-item"
    data-rr-ui-event-key={null}
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    men's clothing
  </div>
  <div
    className="list-group-item"
    data-rr-ui-event-key={null}
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    women's clothing
  </div>
</ul>
`;
